### PR TITLE
Build wireframe transaction framing codec

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implement a wireframe codec for Hotline transaction framing, including 20-byte header handling and multi-fragment payload reassembly
- Add header validation rules and continuation-fragment checks to ensure protocol compliance
- Extend test utilities and add unit, behaviour (BDD), and property tests for the codec
- Wireframe module now exposes the new codec, with accompanying documentation and roadmap updates

### Core
- New module: src/wireframe/codec.rs
  - Implements BorrowDecode for HotlineTransaction to decode the 20-byte header and reassemble fragmented payloads per docs/protocol.md
  - Header validation: ensures flags == 0, total_size <= MAX_PAYLOAD_SIZE (1 MiB), data_size <= MAX_FRAME_DATA (32 KiB), data_size <= total_size, and invalid empty-data-with-nonzero-total cases are rejected
  - Multi-fragment reassembly: validates continuation headers (same flags, is_reply, type, id, error, total_size), enforces remaining-byte constraints, and concatenates payload chunks
  - HotlineTransaction accessor methods: header(), payload(), and into_parts()
- src/wireframe/mod.rs
  - Expose the new codec module publicly: add `pub mod codec;`
- src/transaction.rs
  - Publicly expose `validate_payload(tx: &Transaction)` to support codec testing and validation flows

### Tests & Helpers
- src/wireframe/test_helpers.rs
  - Added transaction_bytes(header, payload) to serialize a single frame (20-byte header + payload)
  - Added fragmented_transaction_bytes(header, payload, fragment_size) to build multi-fragment frames and assemble into a single stream for testing
- tests/features/wireframe_transaction.feature
  - New BDD feature describing wireframe transaction decoding scenarios (single frame, empty, multi-fragment, invalid length, invalid flags, and size limits)
- tests/wireframe_transaction.rs
  - Behavioural and property tests for the wireframe transaction codec
  - Unit tests cover valid single-frame, empty-frame, invalid length combinations, invalid flags, and size-limit errors
  - Property tests (via proptest) validate roundtrip for single-frame, multi-fragment reassembly, and rejection of invalid length combinations

### Documentation & Roadmap
- docs/design.md
  - Added a dedicated section describing the Transaction framing codec (header validation, multi-fragment reassembly, and testing strategy) and how HotlineTransaction is produced from decoding
- docs/roadmap.md
  - Marked the wireframe framing task as done with a note for the completion date

### Dependencies
- Cargo.toml
  - Added proptest = "1.4" for property-based testing (dev/dependencies)

## Testing plan
- [x] Unit tests for header validation and fragment handling in src/wireframe/codec.rs
- [x] Behavioural tests via BDD scenarios in tests/features/wireframe_transaction.feature and tests/wireframe_transaction.rs
- [x] Property tests using proptest to verify valid roundtrips and rejection of invalid length combinations
- [x] End-to-end integration of test helpers to construct frames and fragments for decoding

## How to run
- cargo test
- cargo test --features ... (no special features required; dev-deps used for tests)

## Notes
- This change introduces a new wireframe codec that centralizes header validation and fragment reassembly during decoding, aligning with documentation in docs/protocol.md. It is designed to be exercised through unit tests, BDD scenarios, and property-based tests to ensure robust handling of fragmentation and edge cases.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---
ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/a1a227d1-475c-42a6-98fa-36b60a922a09

## Summary by Sourcery

Introduce a wireframe transaction decoding codec that validates Hotline transaction headers and reassembles fragmented payloads, along with supporting helpers, tests, and documentation updates.

New Features:
- Add a HotlineTransaction wireframe codec that decodes 20-byte transaction headers and reassembles multi-fragment payloads into a single transaction object.

Enhancements:
- Expose the wireframe codec module from the wireframe namespace for external use.
- Make transaction payload validation publicly available to support codec and validation workflows.
- Add helper functions to build single-frame and fragmented transaction byte streams for tests and future integration.

Build:
- Add the proptest dev-dependency to support property-based testing of the wireframe transaction codec.

Documentation:
- Document the transaction framing codec design, including header validation rules, multi-fragment reassembly behaviour, and testing strategy in the design documentation.
- Update the roadmap to mark the wireframe transaction framing codec task as completed with its completion date.

Tests:
- Add BDD feature scenarios describing wireframe transaction decoding, including valid single-frame, empty, multi-fragment, and invalid header/size cases.
- Add behavioural, unit, and property-based tests for the wireframe transaction codec, covering roundtrip decoding, fragmentation reassembly, and rejection of invalid length combinations.
</body>

### CODE REVIEW

File: tests/wireframe_transaction.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        let cfg = config::standard()
+            .with_big_endian()
+            .with_fixed_int_encoding();
+        let result = borrow_decode_from_slice::<HotlineTransaction, _>(&bytes, cfg);
+
+        prop_assert!(result.is_ok(), "decode failed: {:?}", result.err());
+        let (tx, _) = result.unwrap();
+        prop_assert_eq!(tx.header().total_size, total_u32);
+        prop_assert_eq!(tx.payload().len(), total_size);
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/wireframe_transaction.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `fragmented_transaction_bytes`, consider asserting (e.g. via `debug_assert_eq!`) that `header.total_size` matches `payload.len()` to catch mismatches early since the docs require that invariant.
- The bincode config (`config::standard().with_big_endian().with_fixed_int_encoding()`) is duplicated several times in `tests/wireframe_transaction.rs`; refactoring this into a small helper function or constant would reduce repetition and the chance of configuration drift.
- You’ve exposed `transaction::validate_payload` but the new codec never calls it; if the intent is to ensure decoded `HotlineTransaction` payloads are structurally valid, consider wiring this validation into the decode path (or leaving a comment explaining why it’s intentionally omitted).

## Individual Comments

### Comment 1
<location> `tests/wireframe_transaction.rs:229-238` </location>
<code_context>
+    /// Multi-fragment transactions reassemble correctly.
</code_context>

<issue_to_address>
**suggestion (testing):** Bias the `multi_fragment_reassembly` property test to exercise multi-fragment cases more reliably

The current generator lets `fragment_size >= total_size`, so this often collapses into a single-fragment case and overlaps with `roundtrip_valid_single_frame`, reducing coverage of true multi-fragment behaviour. Consider constraining `fragment_size < total_size` and possibly enforcing at least two fragments (e.g. `fragment_size in 1..(total_size / 2).max(1)`) so the property more reliably stresses multi-fragment reassembly, especially the final partial fragment.

Suggested implementation:

```rust
    /// Multi-fragment transactions reassemble correctly.
    #[test]
    fn multi_fragment_reassembly(
        ty in 100u16..500u16,
        id in 1u32..u32::MAX,
        total_size in 4usize..65536usize,
        fragment_size in 100usize..MAX_FRAME_DATA,
    ) {
        // Bias towards true multi-fragment cases by discarding inputs where the fragment
        // size would collapse everything into a single frame or a single full fragment.
        //
        // Since `total_size >= 4`, requiring `fragment_size <= total_size / 2` guarantees
        // at least two fragments and increases the chance of exercising a final partial
        // fragment in the reassembly logic.
        proptest::prop_assume!(fragment_size <= total_size / 2);

        let mut payload = vec![0u8; total_size];
        if total_size >= 2 {
            payload[0] = 0;

```

1. If this module does not already import the proptest macros (e.g. via `use proptest::prelude::*;`), you may need to add `use proptest::prop_assume;` or `use proptest::prelude::*;` at the top of the file so `proptest::prop_assume!` is available without qualification, or keep the fully-qualified call as shown.
2. If you want to *strictly* enforce at least two fragments while still allowing larger `fragment_size` when `total_size` is very small, you could further tweak the assumption to something like `proptest::prop_assume!(fragment_size < total_size);` in addition to or instead of the `/ 2` bound, but the `/ 2` constraint already guarantees at least two fragments for the current `total_size` range.
</issue_to_address>

### Comment 2
<location> `docs/design.md:460` </location>
<code_context>
+
+1. **Unit tests** (`src/wireframe/codec.rs`) use `rstest` to cover single-frame
+   and multi-fragment decoding with parametrised test cases.
+2. **BDD scenarios** (`tests/features/wireframe_transaction.feature`) express
+   acceptance criteria in Gherkin syntax, with step definitions in
+   `tests/wireframe_transaction.rs`.
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The acronym “BDD” is introduced without being defined on first use, which violates the requirement to define uncommon acronyms.

Please expand the first occurrence of BDD, for example: `**Behaviour-Driven Development (BDD) scenarios** ...`. Subsequent references can use just “BDD”.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: src/wireframe/codec.rs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        let err = read_preamble::<_, HotlineTransaction>(&mut reader)
+            .await
+            .expect_err("decode must fail");
+
+        assert!(
+            err.to_string().contains("data size exceeds maximum"),
+            "expected 'data size exceeds maximum' in '{err}'"
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/wireframe/codec.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Has the following now been resolved?

❌ Outstanding (1/6)

validate_payload not called in codec: The transaction::validate_payload function was made public but is never invoked in src/wireframe/codec.rs, and there's no comment explaining why payload validation is intentionally omitted from the decode path. You'll need to either:

    Wire validate_payload into the BorrowDecode implementation, or
    Add a comment explaining why structural validation is deferred to command handlers

**Comment by leynos** [comment]
@coderabbitai review
